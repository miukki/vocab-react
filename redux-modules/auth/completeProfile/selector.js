import Validator from '../../../lib/Validator'
// status in boolean

export const selectShow = state => {
  return state.toJS().show
}

export const selectSubmitting = state => {
  return state.toJS().submitting
}

export const selectHasError = state => {
  return Validator.isAnyValidateError(state.toJS().error)
}

// values
export const selectName = state => {
  return state.toJS().name
}

export const selectErrorTypes = state => {
  return Validator.getErrorTypes(state)
}
