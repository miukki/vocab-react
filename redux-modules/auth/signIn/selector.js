import Validator from '../../../lib/Validator'

// status in boolean
export const selectShow = state => {
  return state.toJS().show
}

export const selectSubmitting = state => {
  return state.toJS().submitting
}

export const selectShowPassword = state => {
  return state.toJS().showPassword
}

export const selectHasError = state => {
  return Validator.isAnyValidateError(state.toJS().error)
}

// values
export const selectMobile = state => {
  return state.toJS().mobile
}

export const selectPassword = state => {
  return state.toJS().password
}

export const selectErrorTypes = state => {
  return Validator.getErrorTypes(state)
}
