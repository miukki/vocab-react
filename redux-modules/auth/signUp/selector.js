import Validator from '../../../lib/Validator'

// status in boolean
export const selectShow = state => {
  return state.toJS().show
}

export const selectSubmitting = state => {
  return state.toJS().submitting
}

export const selectVcodeSending = state => {
  return state.toJS().vcodeSending
}

export const selectVcodeSentAt = state => {
  return state.toJS().vcodeSentAt
}

export const selectVcodeDisabled = state => {
  return state.toJS().vcodeCountdown > 0 || selectVcodeSending(state)
}

export const selectShowPassword = state => {
  return state.toJS().showPassword
}

export const selectHasError = state => {
  return Validator.isAnyValidateError(state.toJS().error)
}

// values
export const selectMobile = state => {
  return state.toJS().mobile
}

export const selectVcode = state => {
  return state.toJS().vcode
}

export const selectVcodeCountdown = state => {
  return state.toJS().vcodeCountdown
}

export const selectPassword = state => {
  return state.toJS().password
}

export const selectErrorTypes = state => {
  return Validator.getErrorTypes(state)
}
