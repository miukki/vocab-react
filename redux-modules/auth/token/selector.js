import jwtDecode from 'jwt-decode'

export const selectUser = state => {
  return state.toJS().jwt && jwtDecode(state.toJS().jwt).user
}

export const selectBgImgIndexBySignUpDate = state => {
  const IMAGES_AMOUNT = 50
  const IMAGE_CHANGE_DAYS = 2

  const daysOfLoop = IMAGE_CHANGE_DAYS * IMAGES_AMOUNT

  const oneDay = 24 * 60 * 60 * 1000 // hours*minutes*seconds*milliseconds
  const signUpDate = new Date(
    // have to `.substr(0, 10)` because safari can't parse time string with timezone
    // reference: https://stackoverflow.com/a/6428201/811653
    state.toJS().jwt &&
      jwtDecode(state.toJS().jwt).user.created_at.substr(0, 10)
  )

  // how many days from user signed up
  const diffDays = Number.parseInt(
    Math.abs((signUpDate.getTime() - Date.now()) / oneDay)
  )

  const daysInCurrentLoop =
    diffDays - Number.parseInt(diffDays / daysOfLoop) * daysOfLoop

  // imageIndex start from 0
  const imageIndex = Number.parseInt(daysInCurrentLoop / IMAGE_CHANGE_DAYS)
  return imageIndex
}
