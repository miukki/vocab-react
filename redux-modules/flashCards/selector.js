export const selectItems = state => state.toJS().items
export const selectIndex = state => state.toJS().index
export const selectFlippedItemIndex = state => state.toJS().flippedItemIndex
