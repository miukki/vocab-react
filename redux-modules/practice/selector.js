// process state
export const selectRemain = state => state.toJS().remain
export const selectPlanAmount = state => state.toJS().planAmount
export const selectShowingResult = state => state.toJS().showingResult
export const selectComplete = state => state.toJS().complete

// qustion
export const selectWordStr = state => state.toJS().wordStr
export const selectPartsOfSpeech = state => state.toJS().partsOfSpeech
export const selectPhoneticSymbol = state => state.toJS().phoneticSymbol

// alternatives
export const selectAnswers = state => state.toJS().answers

// user choice
export const selectSelected = state => state.toJS().selected
