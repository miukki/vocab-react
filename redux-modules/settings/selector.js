export const selectLocation = state => {
  return state.toJS().location
}
