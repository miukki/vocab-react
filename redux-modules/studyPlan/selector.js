import models from 'eic-vocab-app-models'
import wordsLists, { all as allWords } from '../../data/words_lists'

export const selectOpts = state => {
  let { learnedWords } = state.toJS()
  const wordsList = selectWordsList(state)

  if (!wordsList) {
    return null
  }

  const wordsAmount = models.StudyPlan.getWordsAmount(
    wordsList,
    Array.from(learnedWords)
  )

  return models.StudyPlan.getOpts(wordsAmount)
}

export const selectWordsList = state => {
  const wordsListUid = selectWordsListUid(state)
  return Array.from(wordsLists.get(wordsListUid))
}

// picker choices data

export const selectDailyWordsAmounts = state =>
  Array.from(selectOpts(state).keys())

export const selectDaysAmounts = state =>
  selectOpts(state)
    ? Array.from(new Set(selectOpts(state).values()).add(1))
    : []

// picker selected value

export const selectPickerDailyWordsAmount = state =>
  state.toJS().picker.dailyWordsAmount

export const selectPickerDaysAmount = state => {
  const opts = selectOpts(state)
  const dailyWordsAmount = selectPickerDailyWordsAmount(state)
  if (!opts) {
    return NaN
  }
  return models.StudyPlan.getDaysAmount(selectOpts(state), dailyWordsAmount)
}

// panel info

export const selectDailyWordsAmount = state => {
  return (
    state.toJS().dailyWordsAmountOfWordsList[selectWordsListUid(state)] || 15
  )
}

export const selectWordsListUid = state => {
  let { wordsListUid } = state.toJS()

  if (Array.from(wordsLists.keys()).indexOf(wordsListUid) === -1) {
    return Array.from(wordsLists.keys())[0] // get first by default
  }

  return wordsListUid
}

export const selectLearnedWords = state =>
  state.toJS().learnedWords ? Array.from(state.toJS().learnedWords) : []

export const selectLearnedWordsInWordsList = state => {
  const wordsList = selectWordsList(state)
  return selectLearnedWords(state).filter(w => {
    return (
      Array.from(wordsList)
        .map(w => w.wordStr)
        .indexOf(w) !== -1
    )
  })
}

export const selectCheckinLogs = state =>
  state.toJS().checkinLogs
    ? Array.from(
        new Set(
          Array.from(state.toJS().checkinLogs)
            .filter(log => log)
            .map(log => log.substr(0, 8))
        )
      )
    : []

export const selectWordsListAmount = state => {
  const wordsList = selectWordsList(state)
  if (!wordsList) {
    return 0
  }
  return wordsList.length
}

export const selectDaysLeft = state => {
  const opts = selectOpts(state)
  const dailyWordsAmount = selectDailyWordsAmount(state)
  if (!opts) {
    return NaN
  }
  const daysLeft = models.StudyPlan.getDaysAmount(
    selectOpts(state),
    dailyWordsAmount
  )
  return daysLeft ? daysLeft : 1
}

export const selectShowDaysPicker = state => state.toJS().showDaysPicker

export const selectLearnedWordObjects = state => {
  const learnedWords = selectLearnedWords(state)
  return allWords.filter(word => {
    return learnedWords.indexOf(word.wordStr) !== -1
  })
}

export const selecteDailyGoalComplete = state => {
  return selectLearnedWordsOfCompletePractice(state) !== null
}

export const selectLearnedWordsOfCompletePractice = state => {
  const checkinLogs = state.toJS().checkinLogs
  const wordsListUid = state.toJS().wordsListUid
  const dateStringOfToday = new Date()
    .toISOString()
    .slice(0, 10)
    .replace(/-/g, '')
  const searchString = dateStringOfToday + '$' + wordsListUid
  const found = Array.from(checkinLogs)
    .filter(log => log)
    .filter(log => log.indexOf(searchString) !== -1)[0]

  if (!found) {
    return null // checkin log not found
  }

  let [date, listName, wordsAmount] = found.split('$')
  return Number.parseInt(wordsAmount)
}

export const selectWordsListComplete = state =>
  selectLearnedWordsInWordsList(state).length === selectWordsListAmount(state)
