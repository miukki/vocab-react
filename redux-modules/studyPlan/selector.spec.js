import { fromJS } from 'immutable'
import * as selector from './selector'

describe('studyPlan selector', function() {
  describe('selectDailyWordsAmount', function() {
    it('select dailyWordsAmount', function() {
      expect(
        selector.selectDailyWordsAmount(
          fromJS({
            wordsListUid: 'IELTS Listening',
            dailyWordsAmountOfWordsList: { 'IELTS Listening': 20 }
          })
        )
      ).toBe(20)
    })

    it('return default value when dailyWordsAmount not found in state', function() {
      expect(
        selector.selectDailyWordsAmount(
          fromJS({
            wordsListUid: 'IELTS Speaking',
            dailyWordsAmountOfWordsList: { 'IELTS Listening': 20 }
          })
        )
      ).toBe(15)
    })
  })

  describe('selectWordsListUid', function() {
    it('select wordsListUid', function() {
      expect(
        selector.selectWordsListUid(fromJS({ wordsListUid: 'IELTS Listening' }))
      ).toBe('IELTS Listening')
    })

    it('return default one when given wordsListUid do not exist', function() {
      expect(
        selector.selectWordsListUid(fromJS({ wordsListUid: 'not exists' }))
      ).toBe('IELTS Listening')
    })
  })

  describe('selectLearnedWordsOfCompletePractice', function() {
    it('select selectLearnedWordsOfCompletePractice', function() {
      Date.prototype.toISOString = jest
        .fn(Date.prototype.toISOString)
        .mockImplementation(() => '2018-05-30 03:59:35 +0800')

      expect(
        selector.selectLearnedWordsOfCompletePractice(
          fromJS({
            wordsListUid: 'IELTS Listening',
            checkinLogs: ['20180530$IELTS Listening$15']
          })
        )
      ).toBe(15)
    })
  })

  describe('selecteDailyGoalComplete', function() {
    it('select selecteDailyGoalComplete', function() {
      Date.prototype.toISOString = jest
        .fn(Date.prototype.toISOString)
        .mockImplementation(() => '2018-05-30 03:59:35 +0800')

      expect(
        selector.selecteDailyGoalComplete(
          fromJS({
            wordsListUid: 'IELTS Listening',
            checkinLogs: ['20180530$IELTS Listening$15']
          })
        )
      ).toBeTruthy()

      expect(
        selector.selecteDailyGoalComplete(
          fromJS({
            wordsListUid: 'IELTS Listening',
            checkinLogs: ['20180501$IELTS Listening$15']
          })
        )
      ).toBeFalsy()

      expect(
        selector.selecteDailyGoalComplete(
          fromJS({
            wordsListUid: 'IELTS Listening',
            checkinLogs: ['20180530$IELTS Reading$15']
          })
        )
      ).toBeFalsy()
    })
  })

  describe('selectCheckinLogs', function() {
    it('select selectCheckinLogs', function() {
      expect(
        selector.selectCheckinLogs(
          fromJS({
            wordsListUid: 'IELTS Listening',
            checkinLogs: [
              '20180530$IELTS Listening$15',
              '20180530$IELTS Reading$15'
            ]
          })
        )
      ).toMatchSnapshot()

      expect(
        selector.selectCheckinLogs(
          fromJS({
            wordsListUid: 'IELTS Listening',
            checkinLogs: [
              '20180501$IELTS Listening$15',
              '20180501$IELTS Reading$15',
              '20180530$IELTS Listening$15',
              '20180530$IELTS Reading$15'
            ]
          })
        )
      ).toMatchSnapshot()
    })
  })
})
