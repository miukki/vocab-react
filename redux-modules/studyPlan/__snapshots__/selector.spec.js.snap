// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`studyPlan selector selectCheckinLogs select selectCheckinLogs 1`] = `
Array [
  "20180530",
]
`;

exports[`studyPlan selector selectCheckinLogs select selectCheckinLogs 2`] = `
Array [
  "20180501",
  "20180530",
]
`;
